sparse_strips/vello_common/src/tile.rs

```diff
@@ -0,0 +1,509 @@
+// Copyright 2025 the Vello Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Primitives for creating tiles.
+
+use crate::flatten::{Line, Point};
+
+/// The width of a tile.
+pub const TILE_WIDTH: u32 = 4;
+/// The height of a tile.
+pub const TILE_HEIGHT: u32 = 4;
+const TILE_WIDTH_SCALE: f32 = TILE_WIDTH as f32;
+const TILE_HEIGHT_SCALE: f32 = TILE_HEIGHT as f32;
+const INV_TILE_WIDTH_SCALE: f32 = 1.0 / TILE_WIDTH_SCALE;
+const INV_TILE_HEIGHT_SCALE: f32 = 1.0 / TILE_HEIGHT_SCALE;
+// The value of 8192.0 is mainly chosen for compatibility with the old cpu-sparse
+// implementation, where we scaled to u16.
+const NUDGE_FACTOR: f32 = 1.0 / 8192.0;
+const SCALED_X_NUDGE_FACTOR: f32 = 1.0 / (8192.0 * TILE_WIDTH_SCALE);
+
+/// A tile represents an aligned area on the pixmap, used to subdivide the viewport into sub-areas
+/// (currently 4x4) and analyze line intersections inside each such area.
+///
+/// Keep in mind that it is possible to have multiple tiles with the same index,
+/// namely if we have multiple lines crossing the same 4x4 area!
+#[derive(Debug, Clone)]
+pub struct Tile {
+    /// The index of the tile in the x direction.
+    pub x: i32,
+    /// The index of the tile in the y direction.
+    pub y: u16,
+    /// The start point of the line in that tile.
+    pub p0: Point,
+    /// The end point of the line in that tile.
+    pub p1: Point,
+}
+
+impl Tile {
+    /// Create a new tile.
+    pub fn new(x: i32, y: u16, p0: Point, p1: Point) -> Self {
+        Self {
+            // We don't need to store the exact negative location, just that it is negative,
+            // so that the winding number calculation is correct.
+            x: x.max(-1),
+            y,
+            p0,
+            p1,
+        }
+    }
+
+    /// Check whether two tiles are at the same location.
+    pub fn same_loc(&self, other: &Self) -> bool {
+        self.x == other.x && self.same_row(other)
+    }
+
+    /// Check whether two tiles are on the same strip.
+    pub fn same_strip(&self, other: &Self) -> bool {
+        self.same_row(other) && (other.x - self.x).abs() <= 1
+    }
```

Comment:
This name could be confusing, as it's only checking if two tiles are adjacent on the same row. To know whether they're on the same strip requires knowing which tiles are in between these two.

It could be rewritten (and dropping the `abs` for a very slight performance improvement) as

```suggestion
    /// Check whether `self` is the previous tile on the same row as `other`, i.e., it is adjacent to the left of `other`.
    pub fn prev_loc(&self, other: &Self) -> bool {
        self.same_row(other) && self.x + 1 == other.x
    }
```

Reply:
I'm not sure I follow here, because if two tiles are adjacent to each other then they automatically will be in the same strip, I think? At least in the current code.

Reply:
Ah, you mean two tiles could technically be on the same strip if they are not adjacent, but in this case the method would return false? That makes sense I suppose.

Reply:
Indeed, I'm bikeshedding the name of the method, as it confused me a little bit.

Reply:
Indeed, I'm bikeshedding the name of the method, as it confused me a little bit. The functionality is fine, but a name indicating it's just checking for tile adjacency would be good.